web/profiles/joinup/src/Access/CanonicalRouteRestrict.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\joinup\Access;
+
+use Drupal\Core\Access\AccessResult;
+use Drupal\Core\Access\AccessResultInterface;
+use Drupal\Core\Routing\Access\AccessInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\rdf_entity\RdfInterface;
+
+/**
+ * Restricts canonical routes to UID 1 only.
+ */
+class CanonicalRouteRestrict implements AccessInterface {
+
+  /**
+   * Constructs a CanonicalRouteRestrict object.
+   *
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The route match object to be checked.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The account being checked.
+   *
+   * @return \Drupal\Core\Access\AccessResultInterface
+   *   The access result.
+   */
+  public function access(RouteMatchInterface $route_match, AccountInterface $account): AccessResultInterface {
+    $rdf_entity = $route_match->getParameter('rdf_entity');
+    if (!($rdf_entity instanceof RdfInterface) || $rdf_entity->bundle() !== 'spdx_licence') {
```

Comment:
I can see the benefit of having a generic entity canonical route blocker. Unfortunately, we break this promise in this line because we hardcode here the entity type and the bundle. Sad. With his hardcoding we should move the whole logic in the `joinup_licence` module as is no more a _generic solution_.

However, this has challenged me. What if...

* In `\Drupal\joinup\Routing\RouteSubscriber::alterRoutes()`, instead of hardcoding the requirement, we use this pattern: `'_canonical_route_restrict__' . $entity_type_id`. Where `$entity_type_id` is to be computed from the route name or other route data (didn't investigated).
* The service `joinup.canonical_route_restrict` is moved into `joinup_licence` module as:
  ```yaml
   joinup_licence.spdx_entity_route_restrict:
   class: Drupal\joinup\Access\CanonicalRouteRestrict
   arguments: ['@current_route_match', '@current_user', rdf_entity, spdx_licence]
   tags:
     - { name: access_check, applies_to: _canonical_route_restrict__rdf_entity }
   ``` 
* The access method would look like:
  ```php
  public function access(RouteMatchInterface $route_match, AccountInterface $account, string $entity_type_id, string $bundle === NULL) {
    $entity = $route_match->getParameter($entity_type_id);
    if ($entity instanceof ContentEntityInterface && ($bundle === NULL || $bundle === $entity->bundle()) {
       return ....restricted
    }
    ...
  }
  ```

Maybe is a crazy idea, I don't know if really works but it would keep the solution generic. The alternative is to move the whole logic in `joinup_licence`.

Reply:
Let's move this Access class to the `joinup_licence` module. It gets confusing to expand this like this because the main problem that this was built like this is that we don't get the parameters (like an actual rdf_entity entity object) in the route subscriber - since it works directly on routes and not paths.
Also, I will have to move the assignment of the route requirement to the licence so that it is independent as a module.